organization/_posts/2018-03-21-packaging.md

```diff
@@ -0,0 +1,132 @@
+- Examples given for installing Nix in Docker images, noted current limitations
+  - Binary downloads are slow
+  - Package signatures aren’t checked yet
+  - Some packages have issues when built inside Docker images (test suites that require privileges)
+- Why use nix?
+  - Built software should run on “any” Linux flavour (own libc, "build almost everything" approach)
+  - Simpler environments
+  - Many packages already upstream in NixPkgs
+  - Adding new packages is straight forward (many things like shbang patching, RUNPATH etc automagically handled)
+  - Active community
+- What isn’t so good
+  - Documentation lacking, but is improving and many examples help
+  - Nix language has a steep learning curve, but not so bad if just writing expressions
+  - Some issues with independence from host system
+    - Graphics, kernel (but no worse that others)
+  - Sometimes reproducible builds aren’t reproducible (mostly due to remote file URLs being changed)
+- Giulio
+  - Not easy to use “system” or “private” pieces, e.g., a local compiler
+  - Rebuild time is very long, if something changes
+    - Incremental build use case
+  - How to do my day to day development work?
+    - Especially changing some mid-level component, e.g. ROOT but checking impact on AliPhysics (without rebuilding everything?)
+      - AliBuild takes care of this
+    - Chris - Hydra will allow you to download a bootstrap script that gives a local build environment, that could be tweaked locally
+- Marco
+  - User development should be in nixshell
+  - At the moment we are trapped on a single well controlled environment (e.g., lxplus) as we can’t really support non-native systems
+  - Single tool to build and give a development environment is a big plus
+- Pere
+  - To build Gaudi source an lcgview, then cmake && make - there is nothing easier
+  - We are schizophrenic - we want perfect reproducibility, but also want to be able to patch things (e.g., new xrootd client)
+  - How do we solve that problem?
+  - Ben: lifecycle of different pieces is very different, physics makes numbers, middleware is a different matter (and breaks, can need changed)
+  - This has pushed use from the system, instead of building everything
+  - In nix can rebuild xrootd, but with new version?
+- Marco - don’t like to have openssl, etc. in our stack as we are not in charge of updating them
+  - How to install xrootd without recompiling everything?
+  - Divide package sets in two and don’t introduce unnecessary dependencies (careful package management)
+    - Setup a runtime that separates ROOT with ROOT xrootd plugin
+    - Only rebuild pieces needed, recreate views
+    - Does this actually work?
+In favour of building everything, once, instead of having to support patching on many platforms.
+- Guilherme - change openssl, it rebuilds, but only for packages with a direct dependency, otherwise not
+- Can nix use system packages? No.
```

Comment:
This is not completely true. Nix can use system libraries on mac (e.g. all the Cocoa stuff) and I think it can be tricked in doing so also on linux, albeit I cannot find any relevant link about that right now.

Reply:
OK, them maybe we should just note this may be possible, but it's not the default and would require some investigation